nixos/modules/services/web-apps/bookstack.nix

```diff
@@ -0,0 +1,320 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.bookstack;
+  bookstack = cfg.package;
+  db = cfg.database;
+  mail = cfg.mail;
+
+  user = cfg.user;
+  group = config.services.nginx.group;
+
+in {
+  options.services.bookstack = {
+
+    enable = mkEnableOption ''
+      Enable BookStack.
+      The default admin details are "admin@admin.com" with a password of "password". You should change these details immediately after logging in for the first time.
+      This will also enable <option>services.phpfpm</option> and <option>services.nginx</option>.
+    '';
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.bookstack.override {
+        dataDir = cfg.dataDir;
+      };
+      defaultText = "pkgs.bookstack";
+      description = "BookStack package to use.";
+    };
+
+    user = mkOption {
+      default = "bookstack";
+      description = "User bookstack runs as.";
+      type = types.str;
+    };
+
+    appKeyFile = mkOption {
+      default = "/run/keys/bookstack-appkey";
+      description = ''
+        A file containing the AppKey.
+        Used for encryption where needed. Can be generated with <code>head -c 32 /dev/urandom| base64</code> and must be prefixed with <literal>base64:</literal>.
+      '';
+      type = types.str;
+    };
+
+    appURL = mkOption {
+      description = ''
+        The root URL that you want to host BookStack on. All URLs in BookStack will be generated using this value.
+        If you change this in the future you may need to run a command to update stored URLs in the database. Command example: php artisan bookstack:update-url https://old.example.com https://new.example.com
```

Comment:
You currently place the `.env` file in the `stateDir` instead of the root of the package, which means that `php artisan` commands won't work - I believe.

Reply:
I've put following line into the package which symlinks that path and for me `php artisan` commands work.
```
ln -s ${dataDir}/.env $out/.env
```

Reply:
If you run `php artisan config:cache` with either a `.env` file in place, or simply by setting environment variables, these values will be cached under `/bootstrap/cache/config.php` and the `.env` file won't be read anymore... so it isn't clear exactly the best approach here (two obvious options, one of which you have taken).

I'll think about this some more, but what you are doing is fine as long as you protect the permissions on that file :+1:

Reply:
I have thought about it and what you are doing with the `.env` symlink is the best solution :+1: